dlib/constexpr_if.h

```diff
@@ -0,0 +1,131 @@
+// Copyright (C) 2022  Davis E. King (davis@dlib.net)
+// License: Boost Software License   See LICENSE.txt for the full license.
+#ifndef DLIB_IF_CONSTEXPR_H
+#define DLIB_IF_CONSTEXPR_H
+
+#include <utility>
+
+namespace dlib
```

Comment:
This stuff needs docs of some kind

Reply:
I'm tempted to say we shouldn't put this in the public API. It's more of a utility for implementation details.

Reply:
There are many ways you can misuse this, particularly how to use the underscore function properly, which is critical.

Reply:
Na, totally needs a clean and clear interface.  Good software is made in layers where each layer is written with a clear interface.  Like when I'm looking at this I don't know how to use it.  I see what the tests are doing but still.

Reply:
Like why is that `_` function needed?  I've been looking at this and I even just tested some of it and I still don't see why that is needed.  I haven't even found a case where something bad happens without it yet.

Reply:
If you look at try_invoke() in the unit tests you'll see why it's needed. Otherwise I'll upload code to demonstrate Monday.

Reply:
I'm fairly certain if you don't use the _ function in try_invoke the program won't compile

Reply:
Works fine without `_` there too :shrug:

Reply:
Weird. I don't have my PC. I'll check Monday. I mean I would rather not use it but I swear there were circumstances where it was required.

Reply:
Excellent news if it's never required

Reply:
Originally I was getting the same problem described in boost::Hana. Dionne explains the following: "Here, we capture the enclosing args... by reference using [&], and we do not need to receive any arguments. Also, hana::eval_if assumes that its arguments are branches that can be called, and it will take care of calling the branch that is selected by the condition. However, this will still cause a compilation failure, because the bodies of the lambdas are not dependent anymore, and semantic analysis will be done for both branches even though only one would end up being used. The solution to this problem is to make the bodies of the lambdas artificially dependent on something, to prevent the compiler from being able to perform semantic analysis before the lambda is actually used. To make this possible, hana::eval_if will call the selected branch with an identity function (a function that returns its argument unchanged (_ function)), if the branch accepts such an argument". Now it's possible using the function overloading technique it's no longer required. I'll check Monday. I'll clarify though, the original implementation wasn't taken from boost. It's just I was seeing the same problems. Hence why I needed the _ identify function

Reply:
I don't know how you made it work. I've committed the change where `_` isn't used, and everything breaks. I imagine you were forwarding all the arguments as opposed to capturing everything by reference in the lambdas? If so, we were comparing apples and oranges.